Reconfigure screens upon screen_change events

m-col opened the pull request:
This makes Qtile internally respond to changes in physical monitor setup
by reconfiguring screens, making it unnecessary to fully restart Qtile
to match the setup. Fixes #1874.

m-col commented:
Correction - this does not _fix_ #1874, as that would be resolved by completely re-configuring groups across restarts rather than keeping around groups that aren't in the config.

m-col commented:
Should we merge this? I've been using it since I posted it here and it works really well. Anyone else tested it?

m-col commented:
I tell a lie - I actually have a minor change that I never pushed to github, and I've been using that commit. I'll push that now and then it should be good to go.

m-col commented:
Scratch that - I already pushed it 18 days ago...

tcld commented:
I have a question about this:
On my setup I frequently swap out display one with a projector; both have the same resolution, so after a while of using Qtile I noticed that I didn't require a restart at all and everything would still work fine after a switch. With this PR, however, Qtile would react to the switch in some manner. Do you have an idea what that would entail? (If you rebase this, I may give it a go on my system.)

I took a look at the code, by the way. Looks good to me, if that counts for anything. :+1:

m-col commented:
Thanks for the heads up, just rebased.

The `screen_change` hook is fired when there is an change in the geometry of the X screen. This is more of a software thing, i.e. if you enable/disable/reposition external monitors using xrandr/arandr, then it is fired. It is _not_ fired when there is only a hardware change, so your disconnecting and reconnecting would be unaffected if you aren't touching the randr configuration.

Please do try it! I think it would be really valuable, as it would start to move qtile restarts into the realm of "only when you change configuration", which would help close a few open issues.

tcld commented:
I'll check this out immediately. Any suggestions what I should test? I can swap monitors, and maybe move screens in arandr a little. But that's all I can come up with.

m-col commented:
I guess that is all really, as well as enabling/disabling them of course. None
of it should require a restart, and everything should continue to work as
normal.

Some things might need manual setup to work nicely though. For example, I
configure 2 `Screens`s each their a bar, each with a GroupBox widget. So when I
have 1 screen I want the first widget to show all of my groups (6 of them),
whereas if I have 2 screns then I split them 3 and 3 across the 2 screens. The
2nd groupdbox widget always shows the last 3 groups, but it isn't drawn if there
is only 1 screen. 

```
def _refresh_screens():
     if len(qtile.screens) > 1:
         groupboxes[0].visible_groups = ['1', '2', '3']
     else:
         groupboxes[0].visible_groups = ['1', '2', '3', 'q', 'w', 'e']
     groupboxes[0].bar.draw()

hook.subscribe.screen_change(lambda *_: qtile.call_later(0.2, _refresh_screens))
hook.subscribe.startup(lambda *_: qtile.call_later(0.2, _refresh_screens))
```


On 201107 01:33, tcld wrote:
>I'll check this out immediately. Any suggestions what I should test? I can swap monitors, and maybe move screens in arandr a little. But that's all I can come up with.
>
>-- 
>You are receiving this because you authored the thread.
>Reply to this email directly or view it on GitHub:
>https://github.com/qtile/qtile/pull/1899#issuecomment-723497972

tcld commented:
Playing with arandr seems to work pretty well. I moved around screens and swapped the primary and secondary one. Haven't disabled anything, since my config right now sets up two screens and I don't know what Qtile would do if there was only one.

The only issue I have observed is that two of my widgets are kind of drawn on top of each other: `YahooWeather` and `Systray`. I think the latter is partially drawn on top of the former. But I also think that I have observed this behaviour once before, so the cause is likely outside of this PR.

tcld commented:
I noticed one more thing: When moving a screen that displays a window, it seems the window coordinates aren't updated until I focus the window again. (That's for tiled windows, haven't tested floating ones yet.)
Since the screen_rect changed, maybe a `layout_all` should be triggered after changing anything about the screens?

m-col commented:
Ah yeah good idea.

On 201107 02:02, tcld wrote:
>I noticed one more thing: When moving a screen that displays a window, it seems the window coordinates aren't updated until I focus the window again. (That's for tiled windows, haven't tested floating ones yet.)
>Since the screen_rect changed, maybe a `layout_all` should be triggered after changing anything about the screens?
>
>-- 
>You are receiving this because you authored the thread.
>Reply to this email directly or view it on GitHub:
>https://github.com/qtile/qtile/pull/1899#issuecomment-723500645

tcld commented:
`layout_all` fixes the problem for tiled windows. Floating windows aren't moved, however, which can technically leave them entirely off-screen when screens are repositioned. I assume that's a problem with the floating layout rather than your PR, though.

tcld commented:
I've been using this for a few days now. There are some occasions where my screens are re-detected even though I wouldn't want it, however. It seems that randr is very trigger-happy when it comes to swapping displays or launching fullscreen games, which often times leads to my displays switching places. (I have already gotten used to my systray being messed up most of the time.)
Coming from cinnamon, I am very familiar with these exact issues. I was surprised to see that Qtile actually handled these edge cases much better by just doing nothing and assuming nothing.

I am still very much for including this PR, but maybe there is a way to make the functionality a little less intrusive?

m-col commented:
What do you mean your screens are redetected? Qtile (with this PR) doesn't 'do'
anything with randr, only responding to the randr events emitted by the X server
and enabling/disabling Screens accordingly. Your displays 'switching places': is
this the content (e.g. windows, bar), or completely switching so that the cursor
has to go off the opposite edge to what you'd expect to jump to the other
monitor? Also, presumably you didn't get these issues before trying out this
commit, right? I find it very strange you'd get these issues... if that is the
case, could you put something in the screen_change hook on master and see if it
is still being triggered lots when it shouldn't be?

On 201112 11:55, tcld wrote:
>I've been using this for a few days now. There are some occasions where my screens are re-detected even though I wouldn't want it, however. It seems that randr is very trigger-happy when it comes to swapping displays or launching fullscreen games, which often times leads to my displays switching places. (I have already gotten used to my systray being messed up most of the time.)
>Coming from cinnamon, I am very familiar with these exact issues. I was surprised to see that Qtile actually handled these edge cases much better by just doing nothing and assuming nothing.
>
>I am still very much for including this PR, but maybe there is a way to make the functionality a little less intrusive?
>
>-- 
>You are receiving this because you authored the thread.
>Reply to this email directly or view it on GitHub:
>https://github.com/qtile/qtile/pull/1899#issuecomment-726584267

tcld commented:
> only responding to the randr events emitted by the X server
That's what I mean: randr is quite trigger-happy, Qtile is not at fault.

> Your displays 'switching places': is this the content ... or completely switching so that the cursor has to go off the opposite edge ...
The latter.

> presumably you didn't get these issues before trying out this commit
That is true, but only because Qtile didn't react to any of the randr events before.

> could you put something in the screen_change hook on master and see if it is still being triggered lots when it shouldn't be
The assumption being that I see the same amount of events? I would hope so, but what would it tell us?

m-col commented:
> That's what I mean: randr is quite trigger-happy, Qtile is not at fault.

I see.

>The assumption being that I see the same amount of events? I would hope so, but what would it tell us?

Yeah, don't worry about it, that was just me thinking about how to confirm the
above. But I do wonder if useful information could be extract from the events
that are generated. randr shouldn't be trigger-happy. Indeed I do not see the
same. What X version are you using (mine is 1.20.9-2)? Do you have anything like
auto-randr running?

As the main intention of this PR is to remove the need to restart when
reconfiguring monitors, we could move the hook from libqtile into the default
config. This way people who have issues with it can bind it to something else,
like a key combo, rather than upon screen_change firing.

On 201113 03:47, tcld wrote:
>> only responding to the randr events emitted by the X server
>
>That's what I mean: randr is quite trigger-happy, Qtile is not at fault.
>
>> Your displays 'switching places': is this the content ... or completely switching so that the cursor has to go off the opposite edge ...
>
>The latter.
>
>> presumably you didn't get these issues before trying out this commit
>
>That is true, but only because Qtile didn't react to any of the randr events before.
>
>> could you put something in the screen_change hook on master and see if it is still being triggered lots when it shouldn't be
>
>The assumption being that I see the same amount of events? I would hope so, but what would it tell us?
>
>-- 
>You are receiving this because you authored the thread.
>Reply to this email directly or view it on GitHub:
>https://github.com/qtile/qtile/pull/1899#issuecomment-726720813

tcld commented:
Same version of `xorg-server` here. I used autorandr in the past, but it isn't automatically called by anything at the moment. I'll uninstall it and make sure it isn't being used anywhere. Other than that I have xrandr and arandr installed. But they shouldn't act on their own.

Moving the hook might be a good idea, but I also wouldn't want my possibly esoteric use-case to stand in the way of a cool feature.

m-col commented:
I have moved to hook subscription to the default config with a description of what it does, and added a note to the changelog for discoverability. This can help us avoid situtations where people have a trigger-happy randr for some reason!

Beside that, tests are all good, so this is good to go if someone could do a quick review.

tcld commented on libqtile/core/manager.py:
```diff
@@ -360,6 +360,32 @@ def _process_screens(self) -> None:
 
+    def reconfigure_screens(self, ev):
+        """
+        This configures screens upon firing of the screen_change hook.
```
Maybe the comment should be changed or removed, since the connection between `reconfigure_screens` and the mentioned hook is now optional.

m-col replied:
Good call. Done.

tcld commented on libqtile/resources/default_config.py:
```diff
@@ -196,3 +196,10 @@
+# screen change events, for example when reconfiguring monitors with xrandr.
+@hook.subscribe.screen_change
+def reconfigure_screens(e):
+    qtile.reconfigure_screens
```
This doubles as an example of how to use hooks, which is nice. :+1:

The call should probably be `qtile.reconfigure_screens(e)`, or is there some Python magic going on that I don't know about?
Also, would you mind calling the parameter `ev` or even `event`? For me personally, the biggest challenge with Python continues to be finding out what type an object has. More elaborate names help a little.

m-col replied:
Nope, that was me making a dumb typo when I wrapped `qtile.reconfigure_screens` inside the top-level function.

Thanks for the feedback, I've updated the commit.

ramnes commented on libqtile/core/manager.py:
```diff
@@ -361,6 +361,35 @@ def _process_screens(self) -> None:
+
+            scr._configure(self, i, x, y, w, h, self.groups[i])
+            scr.group.layout_all()
+            screens.append(scr)
```
Can we use `_process_screen` here instead of repeating the code?

m-col replied:
I believe so! Just pushed.

ramnes commented on libqtile/core/manager.py:
```diff
@@ -361,6 +361,35 @@ def _process_screens(self) -> None:
+
+        for group in self.groups:
+            if group.screen and group.screen not in screens:
+                group.screen = None
```
Can you tell me more about this? In particular, why are we not assigning the leftovers to one of the new screens?

m-col replied:
This is more for when there are now fewer screens present: any groups that were previously on a screen that is now not there will have have their `screen` reset to `None`. When there are more screens available, then nothing will happen because `group.screen` will be in `screens`.

That does make me wonder -- what happens (on master) when there are more screens than there are groups? I have not accounted for that possibility in this PR.

ramnes replied:
Yeah, my question was badly formulated but that was my concern as well.

Also, I'm thinking that we probably hold a lot more references to the removed screens than just in groups. What should we do with them?

m-col replied:
I just checked all references to screens the other things that save screens directly (rather than accessing via a group or `qtile`) are: `Gap`s/`Bar`s, `widget.Systray`, and `window.Static`. I guess that sounds about right, too.

- Static windows can just be re-positioned to be on the first screen (TODO).
- Gaps and Bars do not need to be handled further; they exist within `qtile.config.screens` and are simply reconfigured when their screen (index) is reconfigured, and this does very little as bars are already set up for multiple calls to `Bar._configure`. Widgets on bars from dormat screens continue to live (and poll) but cannot be accessed until their screen is restored.
- `widget.Systray` _may_ show possibly unexpected behaviour, but I'm not familiar enough with X's systrays to know for sure. I do however use this widget and PR together (and have done for months) and have had no problem. The widget saves an atom telling the server which screen it is on. The screen number doesn't really change (it just disappears) so it will be correct when the screen is restored.

ramnes reviewed (changes requested):
Looks pretty good overall, and most importantly it works well! That's quite simple and still, it's a big improvement for Qtile. :)

m-col commented on libqtile/resources/default_config.py:
```diff
@@ -196,3 +196,10 @@
+# screen change events, for example when reconfiguring monitors with xrandr.
+@hook.subscribe.screen_change
+def reconfigure_screens(event):
+    qtile.reconfigure_screens()
```
👍🏽

m-col commented on libqtile/backend/x11/xcbq.py:
```diff
@@ -779,13 +779,26 @@ def __init__(self, display):
+        self.cursors.finalize()
+        self.disconnect()
+
+    def get_pseudoscreens(self):
```
Yep!

m-col commented on libqtile/backend/x11/xcbq.py:
```diff
@@ -779,13 +779,26 @@ def __init__(self, display):
         extensions = self.extensions()
         self.screens = [Screen(self, i) for i in self.setup.roots]
         self.default_screen = self.screens[self.conn.pref_screen]
+        self.atoms = AtomCache(self)
```
Yep!

ramnes replied:
It's still not where I wanted to try to place it. :sweat_smile: I meant above `self.code_to_syms = {}` (and a blank line), just as in the master branch.

m-col commented:
What do you think about changing the method to `cmd_reconfigure_screens` and exposing it to the command interface?

ramnes commented:
Yeah I think we should have it as a command anyway.

ramnes commented on libqtile/core/manager.py:
```diff
@@ -326,39 +329,66 @@ async def finalize(self):
-            self.screens.append(scr)
+            screens.append(scr)
+
+        return screens
```
Why not modifying `self.screens` directly?

m-col replied:
Hmm it made sense while I was hooking it up with _process_screens but rethinking it, some of these changes might not be necessary.

ramnes commented on libqtile/core/manager.py:
```diff
@@ -326,39 +329,66 @@ async def finalize(self):
-            else:
+            if i < len(self.screens):
+                scr = self.screens[i]
+            elif i < len(self.config.screens):
```
I feel this might be prone to bugs, wouldn't it be safer to use the screen names instead?

m-col replied:
What bug risk do you mean? And screen names?

m-col replied:
The `len(self.screens)` test might be redundant.

ramnes replied:
I'm worried that we use the existing screens on one side of the condition, and config screens on the other side. My brain is quite dead at this time of the day so please bear with me if I'm saying anything dumb, but I have the impression the user change the order of the screens in its configuration, we may end up in a situation where a same Screen object is used for two different physical screens.

m-col replied:
Got it. Using `self.screens` here was redundant anyway, as the same objects are available in the config. Removing it also will prevent that issue you describe :)

ramnes reviewed (approved):
LGTM overall :)

m-col commented:
Alright, I think that's it done now.

`window.Static` will likely need a bit of modification (e.g. giving it a `set_screen` method to change its screen and redraw it), and `Qtile.cmd_reconfigure_screens` will need updating in order to call that method on all static windows that are stuck to now absent screens. However, static windows and their interactions with screens are already broken, but once fixed by #1865, we can make these updates too.

m-col commented:
(The tests have each passed individually during separate runs.)

ramnes commented:
So do you want to wait on #1865 or merge this now? What about the things you listed in https://github.com/qtile/qtile/pull/1899#discussion_r549893128?

m-col commented:
Merge this now. Static windows still work as they always have done and aren't made worse by this, so I don't think we need to wait. The things in the comment are all (generally) accounted for. Bars and Gaps get `_configure`-ed again and work as one would expect. I have no had problems with the systray, but think that it ought to be something to keep an eye out for, rather than something that holds this back, as it may not be an issue. In particular the systray tells the X server which screen it belongs to, but I don't know how/where this is used, and it hasn't been an issue for me running a systray anyway. If an issue does arise, then the first place to look would be to see if that information needs to be accurate at all, if not we can just put screen 0 always.

m-col commented:
I've been thinking about better ways to handle bars that are "removed" than just ignoring them and letting them continue any activity. One option might be to completely tear down bars along with their screen, and then when we are adding screens to read the config again to see if bars need to be (re)created. I think this would be cleaner in that there is no difference in state between having a single screen after removing your second screen, and only starting with a single screen. The main caveat I see with this approach is that the config will be executed during runtime. However, compare that to the current approach to re-configure screens, which is completely restarting Qtile and reading the config anyway.

tych0 commented:
Ping on this, that PR was merged, any chance you can rebase?

m-col commented:
Alrighty, just rebased and fortunately it still works lol.

> I've been thinking about better ways to handle bars that are "removed" than just ignoring them and letting them continue any activity. One option might be to completely tear down bars along with their screen

I think stopping old bars is a good idea, but not re-reading the config. Perhaps we can stop bars but keep them around (in `qtile.config.screens`) in a state where we can start them up again if needed. Bars themselves are already capable of being `_configure`ed multiple times (this is used when adjusting margin sizes to allow for struts), but we'd have to shut down or pause widgets. Perhaps a `_finalize` followed by a `_configure` later already works.

m-col commented:
Actually maybe it's fine to just leave bars to continue running. Looking at the `finalize` methods of the widgets, all it would do is finalise the `Drawers`, which accomplishes nothing really. Arguably _that_ is a waste of the computer's time! So I guess widgets just aren't that much of a burden that shutting them down will be an improvement. If you guys agree then this is ready for review and/or merge :)

tcld commented:
That sounds like the remaining bars and widgets are more of a theoretical problem. Do you know whether they can still count as windows that will then have to be tracked by `core.manager` and thus X? Could they spontaneously show up in unexpected places?

m-col commented:
As it is, these windows will only be visible if they are within a remaining
screen's dimensions. It might be a good idea to simply hide the window then
actually. But doing that, although they would remain as tracked windows, I don't
expect any problems would arise. 

On 210314 11:18, tcld wrote:
>That sounds like the remaining bars and widgets are more of a theoretical problem. Do you know whether they can still count as windows that will then have to be tracked by `core.manager` and thus X? Could they spontaneously show up in unexpected places?
>
>-- 
>You are receiving this because you authored the thread.
>Reply to this email directly or view it on GitHub:
>https://github.com/qtile/qtile/pull/1899#issuecomment-798953737

tcld commented:
It's all edge cases anyway. Probably not worth thinking about overly much until there is a concrete issue that arises because of it.

m-col commented:
Yep, agreed :)

tych0 commented:
Agreed. Let's see what bugs people complain about :)
